pkg/util/pki/parse.go

```diff
@@ -140,3 +140,161 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
 
 	return csr, nil
 }
+
+type PEMBundle struct {
+	CAPEM    []byte
+	ChainPEM []byte
+}
+
+type chainNode struct {
+	cert *x509.Certificate
+
+	issuer *chainNode
+}
+
+// ParseCertificateChainPEM will decode the bundle before passing to
+// ParseCertificateChain
+func ParseCertificateChainPEM(pembundle []byte) (*PEMBundle, error) {
+	certs, err := DecodeX509CertificateChainBytes(pembundle)
+	if err != nil {
+		return nil, err
+	}
+	return ParseCertificateChain(certs)
+}
+
+// ParseCertificateChain will return an x509 PEM encoded certificate chain and
+// CA. The CA may not be a true root, but the highest intermediate certificate.
+// Removes duplicate entries. The returned CA may be empty if a single
+// certificate was passed. Chain contains leaf certificate first. Removes
+// comments and white space. An error is returned if the passed bundle is not a
+// valid flat tree chain, the bundle is malformed, or the chain is broken.
+func ParseCertificateChain(certs []*x509.Certificate) (*PEMBundle, error) {
+	// De-duplicate certificates
+	for i := 0; i < len(certs)-1; i++ {
+		for j := 1; j < len(certs); j++ {
+			if i == j {
+				continue
+			}
+			if certs[i].Equal(certs[j]) {
+				certs = append(certs[:i], certs[i+1:]...)
+			}
+		}
+	}
+
+	var chains []*chainNode
+	// Build the nodes to the linked list
+	for i := range certs {
+		chains = append(chains, &chainNode{cert: certs[i]})
+	}
+
+	for {
+		// We have built the full chain
+		if len(chains) == 1 {
+			break
+		}
+
+		// lastChainsLength is used to ensure that at every pass, the number of
+		// tested chains gets smaller.
+		lastChainsLength := len(chains)
+		for i := 0; i < len(chains)-1; i++ {
+			for j := 1; j < len(chains); j++ {
+				if i == j {
+					continue
+				}
+
+				// attempt to add both chain together
+				chain, ok := chains[i].tryMergeChain(chains[j])
+				if ok {
+					// If adding the chains together was successful, remove inner chain from
+					// list
+					chains = append(chains[:j], chains[j+1:]...)
+				}
+
+				chains[i] = chain
+			}
+		}
+
+		// If no chains were merged in this pass, the chain can never be built
+		// flat. Error.
+		if lastChainsLength == len(chains) {
+			return nil, errors.NewInvalidData("certificate chain is malformed or broken")
+		}
```

Comment:
question: what is the meaning of "flat"?

Reply:
A single tree which has no branches and where each node can have a maximum of two connections to two separate nodes, i.e. a list.

Reply:
It kind of confuses me because "flat tree" is a term already defined in computer science, and anytime I read "flat tree", it makes me think of a tree that has a fixed arity (e.g. binary tree) which allows for storing it as a vector (i.e., a binary tree of depth N can be stored in a vector of length N^2).

I suggest using a different term such as "1-ary tree" or just "list"

Reply:
OK, I will change flat to tree.